src/yb/yql/cql/cqlserver/cql_service.cc

```diff
@@ -229,6 +234,29 @@ void CQLServiceImpl::DeletePreparedStatement(const shared_ptr<const CQLStatement
           << ", memory usage = " << prepared_stmts_mem_tracker_->consumption();
 }
 
+bool CQLServiceImpl::CheckPassword(const std::string plain, const std::string hash) {
+  // FIXME DO NOT MERGE: hash the plaintext password so we're not keeping
+  // multiple plaintext passwords in memory
+
+  {
+    std::lock_guard<std::mutex> guard(password_cache_mutex_);
+    auto cached_entry = password_cache_.get(plain);
+    if (cached_entry && *cached_entry == hash) {
+      return true;
+    }
+  }
+
+  // bcrypt_checkpw has stringcmp semantics
+  bool correct = util::bcrypt_checkpw(plain.c_str(), hash.c_str()) == 0;
```

Comment:
This cache should also serve as a negative cache so we can still get its performance benefits when someone tries a lot of connections with a wrong password:

For example

For these cases let's assume that the username is `foo` and the password is `bar` with a bcypt hash `X` and a sha256 hash `A`. And for password `barbar` we get a sha256 hash`B` and a bcrypt hash `Y`.

Case 1: correct password, not found in the cache.
1. Client sends authentication request with foo/bar.
2. We retrieve password from the master for `foo`, the master returns `X`
3. pswd is checked with `CheckPassword`,  we run sha256 on `bar` and get `A` but find nothing on the cache.
4. We compare bcrypt(`bar`) with the returned value of `X` and they match, so we insert into the catch the pair (`A`, `X`).
5. we return OK because password match

Case 2. Correct password, found in the cache.
Same as above, except on step 3 we find the pair (`A`, `X`) and return OK because passwords match

Case 3. Wrong password, not found in the cache.
1. Client sends authentication request with foo/barbar.
2. We retrieve password from the master for `foo`, the master returns `X`
3. pswd is checked with `CheckPassword`,  we run sha256 on `barbar` and get `B` but find nothing on the cache.
4. We compare bcrypt(`barbar`) with the returned value of `X` and they don't match, but we still insert into the catch the pair (`B`, `Y`).

Case 4. Wrong password, found in the cache
Same as above, but in step 3 we find that the pair (`B`, `Y`) so we don't need to run bcrypt on `barbar`, and we return an error

Reply:
How should we handle passwords being changed while the existing password is in the cache? It's possible that the plaintext password is correct for the new hash, but the old hash is still stored in the cache.